examples/ldns-read-zone.1
.TH ldns-read-zone 1 "30 May 2005"
.SH NAME
ldns-read-zone \- read a zonefile and print it
.SH SYNOPSIS
.B ldns-read-zone 
.IR ZONEFILE 

.SH DESCRIPTION

\fBldns-read-zone\fR reads a DNS zone file and prints it. The output has 1
resource record per line, and no pretty-printing makeup.

.SH OPTIONS
.TP
\fB-0\fR
Print a (null) for the RRSIG inception, expiry and key data. This option
can be used when comparing different signing systems that use the same
DNSKEYs for signing but would have a slightly different timings/jitter.

.TP
\fB-b\fR
Include Bubble Babble encoding of DS's.

.TP
\fB-c\fR
Canonicalize all resource records in the zone before printing

.TP
\fB-d\fR
Only print DNSSEC data from the zone. This option skips every record
that is not of type NSEC, NSEC3 or RRSIG. DNSKEY and DS records are not
printed.

.TP
\fB-e\fR \fIRR type\fR
Do not print RRs of the given \fIrr type\fR.
This option may be given multiple times.
\fB-e\fR is not meant to be used together with \fB-E\fR.

.TP
\fB-E\fR \fIRR type\fR
Print only RRs of the given \fIrr type\fR.
This option may be given multiple times.
\fB-E\fR is not meant to be used together with \fB-e\fR.

.TP
\fB-h\fR
Show usage and exit

.TP
\fB-n\fR
Do not print the SOA record

.TP
\fB-p\fR
Pad the SOA serial number with spaces so the number and the spaces together
take ten characters. This is useful for in file serial number increments.

.TP
\fB-s\fR
Strip DNSSEC data from the zone. This option skips every record
that is of type NSEC, NSEC3 or RRSIG. DNSKEY and DS records are still
printed.

.TP
\fB-S\fR \fI[[+|0]number | YYYYMMDDxx | unixtime ]\fR
Set serial number to the given \fInumber\fR, or when preceded by a sign,
offset the exisiting number with it. When giving the literal strings 
\fIYYYYMMDDxx\fR or \fIunixtime\fR, the serial number is tried to be reset
in datecounter or in unixtime format respectively. Though is the updated serial
number is smaller than the original one, the original one is simply
increased by one.

When updating a serial number, records of type NSEC, NSEC3, RRSIG and DNSKEY
will be skipped when printing the zone.

.TP
\fB-u\fR \fIRR type\fR
Mark \fIRR type\fR for printing in unknown type format.

\fB-u\fR is not meant to be used together with \fB-U\fR.

.TP
\fB-U\fR \fIRR type\fR
Mark \fIRR type\fR for \fBnot\fR printing in unknown type format.

The first occurrence of the \fB-U\fR option marks all RR types for printing
in unknown type format except for the given \fIRR type\fR.
Subsequent \fB-U\fR options will clear the mark for those \fIRR type\fRs too,
so that only the given \fIRR type\fRs will be printed in the presentation 
format specific for those \fIRR type\fRs.

\fB-U\fR is not meant to be used together with \fB-u\fR.

.TP
\fB-v\fR
Show the version and exit

.TP
\fB-z\fR
Sort the zone before printing (this implies -c)


.SH AUTHOR
Written by the ldns team as an example for ldns usage.

.SH REPORTING BUGS
Report bugs to <ldns-team@nlnetlabs.nl>. 

.SH COPYRIGHT
Copyright (C) 2005 NLnet Labs. This is free software. There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.


Changelog
TBD
	* Fix lookup of relative names in ldns_resolver_search.
	* bugfix #548: Double free for answers > 4096 in ldns_resolver_send_pkt
	* Follow CNAME's when tracing with drill (TODO dnssec trace)
	* Fix #551 change Regent to Copyright holder in BSD license in
	  some of the headings of the file, to match the opensource.org
	  BSD license.
	* -e option makes ldns-compare-zones exit with status code 2 on difference
	* Filter out specified RR types with ldns-read-zone -e and -E options

1.6.17	2014-01-10
	* Fix ldns_dnssec_zone_new_frm_fp_l to allow the last parsed line of a
	  zone to be an NSEC3 (or its RRSIG) covering an empty non terminal.
	* Add --disable-dane option to configure and check availability of the
	  for dane needed X509_check_ca function in openssl.
	* bugfix #490: Get rid of type-punned pointer warnings.
	  Thanks Adam Tkac.
	* Make sure executables are linked against libcrypto with the 
	  LIBSSL_LDFLAGS. Thanks Leo Baltus.
	* Miscellaneous prototype fixes. Thanks Dag-Erling Smørgrav.
	* README now shows preferred way to configure for examples and drill.
	* Bind to source address for resolvers. drill binds to source with -I.
	  Thanks Bryan Duff.
	* -T option for ldns-dane that has specific exit status for PKIX
	  validated connections without (secure) TLSA records.
	* Fix b{32,64}_{ntop,pton} detection and handling.
	* New RR type TKEY, but without operational practice.
	* New RR types HIP, NINFO, RKEY, CDS, EUI48, EUI64, URI, CAA and TA.
	* New output format flag (and accompanying functions) to print certain
	  RR's as unknown type
	* -u and -U parameter for ldns-read-zone to mark/unmark a RR type
	  for printing as unknown type
	* bugfix #504: GPOS RR has three rdata fields. Thanks Jelte Jansen.
	* bugfix #497: Properly test for EOF when reading key files with drill.
	* New functions: ldns_pkt_ixfr_request_new and
	  ldns_pkt_ixfr_request_new_frm_str.
	* Use SNI with ldns-dane
	* bugfix #507: ldnsx Fix use of non-existent variables and not
	  properly referring to instance variable.  Patch from shussain.
	* bugfix #508: ldnsx Adding NSEC3PARAM to known/allowable RR type
	  dictionary.  Patch from shussain.
	* bugfix #517: ldns_resolver_new_frm_fp error when invoked using a NULL
	  file pointer.
	* Fix memory leak in contrib/python: ldns_pkt.new_query.
	* Fix buffer overflow in fget_token and bget_token.
	* ldns-verify-zone NSEC3 checking from quadratic to linear performance.
	  Thanks NIC MX (nicmexico.mx)
	* ldns-dane setup new ssl session for each new connect to prevent hangs
	* bugfix #521: drill trace continue on empty non-terminals with NSEC3
	* bugfix #525: Fix documentation of ldns_resolver_set_retry
	* Remove unused LDNS_RDF_TYPE_TSIG and associated functions.
	* Fix ldns_nsec_covers_name for zones with an apex only. Thanks Miek.
	* Configure option to build perl bindings: --with-p5-dns-ldns
	  (DNS::LDNS is a contribution from Erik Ostlyngen)
	* bugfix #527: Move -lssl before -lcrypto when linking
	* Optimize TSIG digest function name comparison (Thanks Marc Buijsman)
	* Compare names case insensitive with ldns_pkt_rr_list_by_name and
	  ldns_pkt_rr_list_by_name_and_type (thanks Johannes Naab)
	* A separate --enable for each draft RR type: --enable-rrtype-ninfo,
	  --enable-rrtype-rkey, --enable-rrtype-cds, --enable-rrtype-uri and
	  --enable-rrtype-ta
	* bugfix #530: Don't sign and verify duplicate RRs (Thanks Jelte Jansen)
	* bugfix #505: Manpage and usage output fixes (Thanks Tomas Hozza)
	* Adjust ldns_sha1() so that the input data is not modified (Thanks
	  Marc Buijsman)
	* Messages to stderr are now off by default and can be reenabled with
	  the --enable-stderr-msgs configure option.

1.6.16	2012-11-13
	* Fix Makefile to build pyldns with BSD make
	* Fix typo in exporting b32_* symbols to make pyldns load again
	* Allow leaving the RR owner name empty in ldns-testns datafiles.
	* Fix fail to create NSEC3 bitmap for empty non-terminal (bug
	  introduced in 1.6.14).

1.6.15	2012-10-25
	* Remove LDNS_STATUS_EXISTS_ERR from ldns/error.h to make ldns
	  binary compatible with earlier releases again.

1.6.14	2012-10-23
	* DANE support (RFC6698), including ldns-dane example tool.
	* Configurable default CA certificate repository for ldns-dane with
	  --with-ca-file=CAFILE and --with-ca-path=CAPATH
	* Configurable default trust anchor with --with-trust-anchor=FILE
	  for drill, ldns-verify-zone and ldns-dane
	* bugfix #474: Define socklen_t when undefined (like in Win32)
	* bugfix #473: Dead code removal and resource leak fix in drill
	* bugfix #471: Let ldns_resolver_push_dnssec_anchor accept DS RR's too.
	* Various bugfixes from code reviews from CZ.NIC and Paul Wouters
	* ldns-notify TSIG option argument checking
	* Let ldns_resolver_nameservers_randomize keep nameservers and rtt's
	  in sync.
	* Let ldns_pkt_push_rr now return false on (memory) errors.
	* Make buffer_export comply to documentation and fix buffer2str
	* Various improvements and fixes of pyldns from Katel Slany
	  now documented in their own Changelog.
	* bugfix: Make ldns_resolver_pop_nameserver clear the array when
	  there was only one.
	* bugfix #459: Remove ldns_symbols and export symbols based on regex
	* bugfix #458: Track all newly created signatures when signing.
	* bugfix #454: Only set -g and -O2 CFLAGS when no CFLAGS was given.
	* bugfix #457: Memory leak fix for ldns_key_new_frm_algorithm.
	* pyldns memory handling fixes and the python3/ldns-signzone.py
	  examples script contribution from Karel Slany.
	* bugfix #450: Base # bytes for P, G and Y (T) on the guaranteed
	  to be bigger (or equal) P in ldns_key_dsa2bin.
	* bugfix #449: Deep free cloned rdf's in ldns_tsig_mac_new.
	* bugfix #448: Copy nameserver value (in stead of reference) of the
	  answering nameserver to the answer packet in ldns_send_buffer, so
	  the original value may be deep freed with the ldns_resolver struct.
	* New -0 option for ldns-read-zone to replace inception, expiration

examples/ldns-read-zone.c
/*
 * read a zone file from disk and prints it, one RR per line
 *
 * (c) NLnetLabs 2005-2008
 *
 * See the file LICENSE for the license
 */

#include "config.h"
#include <unistd.h>
#include <stdlib.h>

#include <ldns/ldns.h>
#include <ldns/host2str.h>

#include <errno.h>

void print_usage(const char* progname)
{
	printf("Usage: %s [OPTIONS] <zonefile>\n", progname);
	printf("\tReads the zonefile and prints it.\n");
	printf("\tThe RR count of the zone is printed to stderr.\n");
	printf("\t-0 zeroize timestamps and signature in RRSIG records.\n");
	printf("\t-b include Bubble Babble encoding of DS's.\n");
	printf("\t-c canonicalize all rrs in the zone.\n");
	printf("\t-d only show DNSSEC data from the zone\n");
	printf("\t-e <rr type>\n");
	printf("\t\tDo not print RRs of the given <rr type>.\n");
	printf("\t\tThis option may be given multiple times.\n");
	printf("\t\t-e is not meant to be used together with -E.\n");
	printf("\t-E <rr type>\n");
	printf("\t\tPrint only RRs of the given <rr type>.\n");
	printf("\t\tThis option may be given multiple times.\n");
	printf("\t\t-E is not meant to be used together with -e.\n");
	printf("\t-h show this text\n");
	printf("\t-n do not print the SOA record\n");
	printf("\t-p prepend SOA serial with spaces so"
		" it takes exactly ten characters.\n");
	printf("\t-s strip DNSSEC data from the zone\n");
	printf("\t-S [[+|-]<number> | YYYYMMDDxx | "
			" unixtime ]\n"
		"\t\tSet serial number to <number> or,"
			" when preceded by a sign,\n"
		"\t\toffset the existing number with "
			"<number>.  With YYYYMMDDxx\n"
		"\t\tthe serial is formatted as a datecounter"
			", and with unixtime as\n"
		"\t\tthe number of seconds since 1-1-1970."
			"  However, on serial\n"
		"\t\tnumber decrease, +1 is used in stead"
			".  (implies -s)\n");
	printf("\t-u <rr type>\n");
	printf("\t\tMark <rr type> for printing in unknown type format.\n");
	printf("\t\tThis option may be given multiple times.\n");
	printf("\t\t-u is not meant to be used together with -U.\n");
	printf("\t-U <rr type>\n");
	printf("\t\tMark <rr type> for not printing in unknown type format.\n");
	printf("\t\tThis option may be given multiple times.\n");
	printf(
	"\t\tThe first occurrence of the -U option marks all RR types for"
	"\n\t\tprinting in unknown type format except for the given <rr type>."
	"\n\t\tSubsequent -U options will clear the mark for those <rr type>s"
	"\n\t\ttoo, so that only the given <rr type>s will be printed in the"
	"\n\t\tpresentation format specific for those <rr type>s.\n");
	printf("\t\t-U is not meant to be used together with -u.\n");
	printf("\t-v shows the version and exits\n");
	printf("\t-z sort the zone (implies -c).\n");
	printf("\nif no file is given standard input is read\n");
	exit(EXIT_SUCCESS);
}

void exclude_type(ldns_rdf **types2show, ldns_rr_type t)
{
	ldns_status s;

	assert(types2show);

	if (! *types2show && LDNS_STATUS_OK !=
			(s = ldns_rdf_bitmap_known_rr_types(types2show)))
		goto fail;

	s =  ldns_nsec_bitmap_clear_type(*types2show, t);
	if (s == LDNS_STATUS_OK)
		return;
fail:
	fprintf(stderr, "Cannot exclude rr type %s: %s\n"
	              , ldns_rr_descript(t)->_name
		      , ldns_get_errorstr_by_id(s));
	exit(EXIT_FAILURE);
}

void include_type(ldns_rdf **types2show, ldns_rr_type t)
{
	ldns_status s;

	assert(types2show);

	if (! *types2show && LDNS_STATUS_OK !=
			(s = ldns_rdf_bitmap_known_rr_types_space(types2show)))
		goto fail;

	s =  ldns_nsec_bitmap_set_type(*types2show, t);
	if (s == LDNS_STATUS_OK)
		return;
fail:
	fprintf(stderr, "Cannot exclude all rr types except %s: %s\n"
	              , ldns_rr_descript(t)->_name
		      , ldns_get_errorstr_by_id(s));
	exit(EXIT_FAILURE);
}

int
main(int argc, char **argv)
{
	char *filename;
	FILE *fp;
	ldns_zone *z;
	int line_nr = 0;
	int c;
	bool canonicalize = false;
	bool sort = false;
	bool print_soa = true;
	ldns_status s;
	size_t i;
	ldns_rr_list *stripped_list;
	ldns_rr *cur_rr;
	ldns_output_format_storage fmt_storage;
	ldns_output_format* fmt = ldns_output_format_init(&fmt_storage);
	ldns_rdf *types2show = NULL;

	ldns_soa_serial_increment_func_t soa_serial_increment_func = NULL;
	int soa_serial_increment_func_data = 0;

        while ((c = getopt(argc, argv, "0bcde:E:hnpsS:u:U:vz")) != -1) {
                switch(c) {
			case '0':
				fmt->flags |= LDNS_FMT_ZEROIZE_RRSIGS;
				break;
			case 'b':
				fmt->flags |= 
					( LDNS_COMMENT_BUBBLEBABBLE |
					  LDNS_COMMENT_FLAGS        );
				break;
                	case 'c':
                		canonicalize = true;
                		break;
                	case 'd':
				include_type(&types2show, LDNS_RR_TYPE_RRSIG);
				include_type(&types2show, LDNS_RR_TYPE_NSEC);
				include_type(&types2show, LDNS_RR_TYPE_NSEC3);
				break;
			case 'e':
				exclude_type(&types2show, 
					ldns_get_rr_type_by_name(optarg));
				break;
			case 'E':
				include_type(&types2show, 
					ldns_get_rr_type_by_name(optarg));
				break;
			case 'h':
				print_usage("ldns-read-zone");
				break;
			case 'n':
				print_soa = false;
				break;
			case 'p':
				fmt->flags |= LDNS_FMT_PAD_SOA_SERIAL;
				break;
			case 's':
			case 'S':
				exclude_type(&types2show, LDNS_RR_TYPE_RRSIG);
				exclude_type(&types2show, LDNS_RR_TYPE_NSEC);
				exclude_type(&types2show, LDNS_RR_TYPE_NSEC3);
				if (c == 's') break;
				if (*optarg == '+' || *optarg == '-') {
					soa_serial_increment_func_data =
						atoi(optarg);
					soa_serial_increment_func =
						ldns_soa_serial_increment_by;
				} else if (! strtok(optarg, "0123456789")) {
					soa_serial_increment_func_data =
						atoi(optarg);
					soa_serial_increment_func =
						ldns_soa_serial_identity;
				} else if (!strcasecmp(optarg, "YYYYMMDDxx")){
					soa_serial_increment_func =
						ldns_soa_serial_datecounter;
				} else if (!strcasecmp(optarg, "unixtime")){
					soa_serial_increment_func =
						ldns_soa_serial_unixtime;
				} else {
					fprintf(stderr, "-S expects a number "
						"optionally preceded by a "
						"+ or - sign to indicate an "
						"offset, or the text YYYYMM"
						"DDxx or unixtime\n");
					exit(EXIT_FAILURE);
				}
				break;
			case 'u':
				s = ldns_output_format_set_type(fmt,
					ldns_get_rr_type_by_name(optarg));
				if (s != LDNS_STATUS_OK) {
					fprintf( stderr
					       , "Cannot set rr type %s "
					         "in output format to "
						 "print as unknown type: %s\n"
					       , ldns_rr_descript(
					       ldns_get_rr_type_by_name(optarg)
						       )->_name
					       , ldns_get_errorstr_by_id(s)
					       );
					exit(EXIT_FAILURE);
				}
				break;
			case 'U':
				s = ldns_output_format_clear_type(fmt,
					ldns_get_rr_type_by_name(optarg));
				if (s != LDNS_STATUS_OK) {
					fprintf( stderr
					       , "Cannot set rr type %s "
					         "in output format to not "
						 "print as unknown type: %s\n"
					       , ldns_rr_descript(
					       ldns_get_rr_type_by_name(optarg)
						       )->_name
					       , ldns_get_errorstr_by_id(s)
					       );
					exit(EXIT_FAILURE);
				}
				break;
			case 'v':
				printf("read zone version %s (ldns version %s)\n", LDNS_VERSION, ldns_version());
				exit(EXIT_SUCCESS);
				break;
                        case 'z':
                		canonicalize = true;
                                sort = true;
                                break;
		}
	}
	argc -= optind;
	argv += optind;

	if (argc == 0) {
		fp = stdin;
	} else {
		filename = argv[0];

		fp = fopen(filename, "r");
		if (!fp) {
			fprintf(stderr, "Unable to open %s: %s\n", filename, strerror(errno));
			exit(EXIT_FAILURE);
		}
	}
	
	s = ldns_zone_new_frm_fp_l(&z, fp, NULL, 0, LDNS_RR_CLASS_IN, &line_nr);

	fclose(fp);
	if (s != LDNS_STATUS_OK) {
		fprintf(stderr, "%s at %d\n", 
				ldns_get_errorstr_by_id(s),
				line_nr);
                exit(EXIT_FAILURE);
	}

	if (types2show) {
		if (print_soa)
			print_soa = ldns_nsec_bitmap_covers_type(types2show,
					LDNS_RR_TYPE_SOA);
		stripped_list = ldns_rr_list_new();
		while ((cur_rr = ldns_rr_list_pop_rr(ldns_zone_rrs(z))))
			if (ldns_nsec_bitmap_covers_type(types2show,
						ldns_rr_get_type(cur_rr)))
				ldns_rr_list_push_rr(stripped_list, cur_rr);
			else
				ldns_rr_free(cur_rr);
		ldns_rr_list_free(ldns_zone_rrs(z));
		ldns_zone_set_rrs(z, stripped_list);
	}

	if (canonicalize) {
		ldns_rr2canonical(ldns_zone_soa(z));
		for (i = 0; i < ldns_rr_list_rr_count(ldns_zone_rrs(z)); i++) {
			ldns_rr2canonical(ldns_rr_list_rr(ldns_zone_rrs(z), i));
		}
	}
	if (sort) {
		ldns_zone_sort(z);
	}

	if (print_soa && ldns_zone_soa(z)) {
		if (soa_serial_increment_func) {
			ldns_rr_soa_increment_func_int(
					ldns_zone_soa(z)
				, soa_serial_increment_func
				, soa_serial_increment_func_data
				);
		}
		ldns_rr_print_fmt(stdout, fmt, ldns_zone_soa(z));
	}
	ldns_rr_list_print_fmt(stdout, fmt, ldns_zone_rrs(z));

	ldns_zone_deep_free(z);

        exit(EXIT_SUCCESS);
}
